### PR TITLE
Product Collection: Add Taxonomy filters to sidebar settings

### LINKED ISSUE #9364: Product Collection - Filters: Taxonomies
### Description
Provide "Taxonomies" filter to the Product Collection block. This setting exists in Products block, so should be reused if possible.

### Goals - acceptance criteria
- Changing the filter should impact the query and the change should be reflected in the Editor's preview and frontend
- Filter should be placed in Filters section
- Taxonomies shows the Categories and Tags filtering options

We can think of providing two separate filters (Categories and Tags), rather than single one called Taxonomies (to be verified in scope of this issue).

### Image
image

### PR DESCRIPTION
**Pull request by imanish003**
> ⚠️ This PR will be merged after PR #9600, as this branch is checked out from the branch of PR #9600.

This PR adds `Taxonomy` filters to sidebar settings.

image
image

Fixes #9364 

### Testing

1. Create a new post
2. Add Product Collection block
3. With `Product Collection` block selected, check the sidebar for the Taxonomy filter i.e., product categories, tags etc. These will be displayed as shown in images below:
image
image

4. Make sure that you have products with categories and tags. If not, add categories or tags to some products.
5. Now select a product category or tags. After the selection, verify that changes reflect immediately in the Editor view.
6. Save and publish the post. After publishing, you can just navigate to the Frontend and verify that the changes you made in the Editor are accurately reflected.
7. Return to the Editor and validate the `Reset all` functionality of filters:
   - Clicking the `Reset all` button resets your selected filters.
   - The `Reset all` button can be found as shown in image below:
image

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add Taxonomy filters to sidebar settings

### CODE REVIEW

File: src/BlockTypes/ProductCollection.php [hunk 1]
```diff
@@ -139,14 +144,17 @@ public function build_query( $query, $block, $page ) {
 			's'              => $block_context_query['search'],
 		);
 
-		$is_on_sale = $block_context_query['woocommerceOnSale'] ?? false;
+		$is_on_sale       = $block_context_query['woocommerceOnSale'] ?? false;
+		$taxonomies_query = $this->get_filter_by_taxonomies_query( $query['tax_query'] );
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php]: I'm getting this PHP warning:

```
Warning: Undefined array key "tax_query" in /wp-content/plugins/woocommerce-blocks/src/BlockTypes/ProductCollection.php` on line 148
```

File: src/BlockTypes/ProductCollection.php [hunk 2]
```diff
@@ -448,4 +456,86 @@ private function merge_tax_queries( ...$queries ) {
...
+			'tax_query' => array_values( $grouped_attributes ),
+		);
+	}
+
+	/**
+	 * Return a query to filter products by taxonomies (product categories, product tags, etc.)
+	 *
+	 * For example:
+	 * User could provide "Product Categories" using "Filters" ToolsPanel available in Inspector Controls.
+	 * We use this function to extract it's query from $tax_query.
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php]: I think there is a typo here: "_We use this function to extract **its** query from $tax_query_".

File: src/BlockTypes/ProductCollection.php [hunk 3]
```diff
@@ -448,4 +456,86 @@ private function merge_tax_queries( ...$queries ) {
...
+	 * Array
+	 *    (
+	 *        [taxonomy] => product_cat
+	 *        [terms] => Array
+	 *            (
+	 *                [0] => 36
+	 *            )
+	 *    )
+	 *
+	 * For product categories, taxonomy would be "product_tag"
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php]: Should this say "_For product **tags**_"?

**Review by Aljullu — COMMENTED** [review]
This is looking great as well, good job @imanish003!

Similar to #9630, _Reset All_ is not working for me.

I also think there should be some more margin between `Separate with commas or the Enter key.` and `Product tags`, otherwise it looks too close:

imatge

Besides that, I left some minor comments inline. :point_down:

File: src/BlockTypes/ProductCollection.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php]: Oh, yes. Good catch! I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509). Just curious, where do you see these warnings? I use the Query Monitor plugin 🙂

File: src/BlockTypes/ProductCollection.php (hunk 2 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php]: I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509)

File: src/BlockTypes/ProductCollection.php (hunk 3 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php]: I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509)

**Comment by imanish003** [comment]
Hey @Aljullu,

> Similar to https://github.com/woocommerce/woocommerce-blocks/pull/9630, Reset All is not working for me.
> I also think there should be some more margin between Separate with commas or the Enter key. and Product tags, otherwise it looks too close:

Fixed in e049cf8. Could you please review the changes once more? I greatly appreciate your valuable feedback. Thank you! 🙌🏻

File: src/BlockTypes/ProductCollection.php [hunk 4]
```diff
@@ -487,7 +490,51 @@ function ( $carry, $item ) {
...
+	 *            )
+	 *    )
+	 *
+	 * For product tags, taxonomy would be "product_tag"
+	 *
+	 * @param array $tax_query Query to filter products by taxonomies.
+	 * @return array Query to filter products by taxonomies.
+	 */
+	private function get_filter_by_taxonomies_query( $tax_query ): array {
+		if ( ! isset( $tax_query ) || ! is_array( $tax_query ) ) {
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php]: Nit: could we simply call `is_array()` and skip the `isset()` check? `is_array` would already return false if `$tax_query` is not set, no? Or do you think the code is clearer this way?

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	/**
+	 * Here we make an extra check for entered terms in a non case sensitive way,
+	 * to match user expectations, due to `FormTokenField` behaviour that shows
+	 * suggestions which are case insensitive.
+	 *
+	 * Although WP tries to discourage users to add terms with the same name (case insensitive),
+	 * it's still possible if you manually change the name, as long as the terms have different slugs.
+	 * In this edge case we always apply the first match from the terms list.
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: TIL, thanks for the explanatory comment. :clap:

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+		}
+		setSuggestions( EMPTY_ARRAY );
+		onChange( Array.from( newTermIds ) );
+	};
+
+	return (
+		<div className="wc-block-editor-product-collection-inspector__taxonomy-control">
+			<FormTokenField
+				label={ taxonomy.name }
+				value={ value }
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: I'm a bit confused by this, but if I'm not wrong, `value` can be directly inferred from `existingTerms`, right? In that case, we shouldn't need to store `value` in the state. Ie, we could do something like this:

```suggestion
				value={ ( existingTerms || [] ).map( ( { id, name } ) => ( {
					id,
					value: name,
				} ) ) }
```

and then get rid of the `state` value and the `useEffect()` associated with it. (I might have missed something, though :sweat_smile:)

**Review by Aljullu — COMMENTED** [review]
Thanks for updating and rebasing this PR, @imanish003! This is looking good, but I found one issue. There is the possibility that two subcategories might have the same name:

imatge

When that happens, there is a JS warning in the console: `Warning: Encountered two children with the same key, Accessories`. And it's impossible for the user to know which category is which.

One idea to fix this could be to append the `slug` when there are two or more terms with the same name, as we know slugs are unique. What do you think? Another idea would be to append the parent category name, but I guess that complicates the code a lot and there is also the possibility that parent categories might have the same name, so it doesn't really solve the problem.

File: src/BlockTypes/ProductCollection.php (hunk 1 above)
Comment (Aljullu) [reply src/BlockTypes/ProductCollection.php]: I see them directly on the screen. I think you need to set these values in `wp-config.php`:

```PHP
define( 'WP_DEBUG', true );
define( 'SCRIPT_DEBUG', true );
```

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx (hunk 5 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: I cannot take credit for this code as I copied it from [here](https://github.com/WordPress/gutenberg/blob/305c93ff4d19cefda9ecd200d2e21b35319b2e6e/packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js). I simply added TypeScript types and made minor improvements. 🙂

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx (hunk 6 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: The reason for using useState is that when a user adds a new term, `existingTerms` becomes empty. Therefore, if we were to use `existingTerms` as you suggested, the input would become empty for a few seconds while the `existingTerms` are being fetched. However, I will take a look at this and see if any improvements can be made. Thanks for pointing this out 🙌🏻

File: src/BlockTypes/ProductCollection.php (hunk 4 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php]: Good catch! I will make the changes. Thanks for pointing this out 🙌🏻

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx (hunk 6 above)
Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: Oooh, you are right, I see what you mean. I wonder if we could use [`usePrevious()`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts), however. IMO it seems like a better hook for this use case. So we would use `existingTerms` by default, but if it's empty, we would use `previousExistingTerms` to make sure the input field doesn't become empty. What do you think? :thinking:

Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: Actually, I did consider using `usePrevious()`, but the implementation would be somewhat complex. For instance, if there is one term selected and the user select another term, the value of `existingTerms` would look like this:

image

As you can see in the console.log output above. There are multiple null and empty values before the second term is added to existing terms. This complicates determining when we should use the `usePrevious` value. Do you see the point? However, I will experiment with it today to see if I can figure out how to use `usePrevious`, which may require some code refactoring.

The first thing I will try is [useEntityRecords](https://developer.wordpress.org/news/2023/05/useentityrecords-an-easier-way-to-fetch-wordpress-data/) hook.

**Comment by imanish003** [comment]
Hey @Aljullu, I have made a good amount of changes:
- Added handling for duplicate term names
- Refactored a lot of code to improve code readability and separation of concerns, thus making future changes and maintenance easier.

Can you please take a look again? 🙂

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+			orderby: 'name',
+			exclude: termIds,
+			per_page: 20,
+		} );
+
+	suggestionsRef.current = useMemo( () => {
+		if ( ! searchHasResolved ) return suggestionsRef.current;
+
+		const newSuggestions = searchResults.map( ( searchResult: Term ) =>
+			termIdToNameMap.get( searchResult.id )
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: On my end, this causes a JS error when `Uncategorized` is one of the results, as it doesn't seem to be present in `termIdToNameMap` for some reason. I don't really know why. Could we add a fallback to `searchResult.name` just in case?

```suggestion
			termIdToNameMap.get( searchResult.id ) || searchResult.name
```

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+};
+
+interface TaxonomyItemProps {
+	taxonomy: Taxonomy;
+	termIds: number[];
+	onChange: ( termIds: number[] ) => void;
+}
+
+// Function to get the term id based on user input in the `FormTokenField`.
+export const getTermIdByTermValue = (
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: Is this function exported for some reason?

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+	id: number;
+	name: string;
+	slug: string;
+};
+
+interface TaxonomyItemProps {
+	taxonomy: Taxonomy;
+	termIds: number[];
+	onChange: ( termIds: number[] ) => void;
+}
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: Nit: maybe a personal taste, but IMO listing TS types _before_ JS constants makes code easier to read as TS types are not part of the file bundle but JS constants are. What do you think?

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	}, [ searchHasResolved, searchResults, termIdToNameMap ] );
+
+	// Update the selected terms when the user selects a suggestion.
+	const onTermsChange = ( newTermValues: FormTokenField.Value[] ) => {
+		const newTermIds = [];
+		for ( const termValue of newTermValues ) {
+			const termId = getTermIdByTermValue(
+				termValue as string | Term,
+				termNameToIdMap
+			);
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:245]: I think that if you check whether `termValue` is a string or not here:

```suggestion
			const termId =
				typeof termValue === 'string'
					? getTermIdByTermValue( termValue, termNameToIdMap )
					: termValue.id;
```

Then, `getTermIdByTermValue()` can be a one-liner:

```JS
const getTermIdByTermValue = (
	searchTerm: string,
	termNameToIdMap: Map< string, number >
): number | undefined => {
	return (
		termNameToIdMap.get( searchTerm ) ||
		termNameToIdMap.get( searchTerm.toLocaleLowerCase() )
	);
};
```

**Review by Aljullu — COMMENTED** [review]
Thanks for updating this PR, @imanish003! I really appreciate how you split the code in two files and now it's much easier to read. I found one issue: the _Uncategorized_ category appears for me in the UI, but I'm not able to select it. Clicking on it doesn't add it to the field. Can you reproduce?

[Enregistrament de pantalla des de 2023-06-01 13-32-55.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/5fc9de40-f747-4072-916c-605940e8808c)

Another thing, which is probably not a blocker, is that I feel this component makes too many calls to the API. Ie: just enabling the "Taxonomies" filter, causes three API calls. It might be nice to investigate if they can be reduced.

[Enregistrament de pantalla des de 2023-06-01 13-47-53.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/5016cd48-210f-4972-b47a-2edf242e3e86)

Besides that, I left some comments inline. Some of them are code suggestions, I did test them but might have missed some edge cases, so feel free to disregard if they don't work well.

**Comment by imanish003** [comment]
Hey @Aljullu 👋 

> I found one issue: the Uncategorized category appears for me in the UI, but I'm not able to select it. Clicking on it doesn't add it to the field. Can you reproduce?

While I was able to add `Uncategorized`, I encountered an issue when trying to add its lowercase version, `uncategorized`. As a result, I pushed a fix in [6a10f3f](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/6a10f3fb5359ab455266c6c273d4cbd9c804f574). Perhaps that was the same problem you faced as well? Could you please check if this fix works for you too?

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 7 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: I can think of only one case when this might happen, i.e. when we are editing Product Collection block in a post and in a new tab we add a new term. In that case, new term won't be available in termIdToNameMap until user refreshes the page. 

Can you think of any other case when this might happen? 

I am adding a fallback for such scenarios. 🙂

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 8 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: My apologies. I accidentally left this function exported while moving it between files. There is no reason for this function to be exported. 😅

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 9 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: I don't have any specific preference regarding this, but I will go ahead and move the types before the constants as you suggested. 😀

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 10 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:245]: I understand your point, but I believe it's better to leave the implementation details to our abstraction, i.e. `getTermIdByTermValue`. I have made some improvements to this function based on your feedback in [d9d7cd7](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/d9d7cd77284801aad67c74a2536fc831642acb1d).

However, please let me know if you prefer checking the type outside of `getTermIdByTermValue`. I can make that change because I don't have a strong opinion on this. 🙂

**Comment by imanish003** [comment]
@Aljullu 
> Another thing, which is probably not a blocker, is that I feel this component makes too many calls to the API. Ie: just enabling the "Taxonomies" filter, causes three API calls. It might be nice to investigate if they can be reduced.

I have implemented the changes in [7b56958](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/7b56958aa1c69e15ebc849707ae79c4131b6a67e). I have optimized the data fetching process, resulting in a single call. Can you please take a look?

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+	searchTerm: Term | string,
+	termNameToIdMap: Map< string, number >
+): number | undefined => {
+	const termId = ( searchTerm as Term )?.id;
+	if ( termId ) {
+		return termId;
+	}
+
+	return (
+		termNameToIdMap.get( searchTerm as string ) ||
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:43]: After 6a10f3fb5359ab455266c6c273d4cbd9c804f574, we don't need this line anymore, no? Because `termNameToIdMap` already includes lower-case terms so we can default to it.

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx (hunk 6 above)
Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx]: Just for future reference: heads-up that `usePrevious()` accepts a [`validation` function](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts#L18) so you can determine whether it updates or not. So in this case, it should be possible to make it so it doesn't update in certain circumstances.

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 7 above)
Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: Hmmm, I'm not sure what was going on. Tested on a different site and I can't reproduce it there. Let's leave it as-is and I will open an issue if I can reproduce again. Sorry for the confusion. :sweat:

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 10 above)
Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:245]: Thanks! I usually prefer functions to accept the minimum possible types as possible, so they are easier to understand, but no need to change it.

**Review by Aljullu — APPROVED** [review]
LGTM, thanks for working on this, @imanish003!

I left one minor suggestion and answered all open threads, but approving. :ship:

File: assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 11 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:43]: I think we need it because of [this reason](https://github.com/WordPress/gutenberg/blob/a993bfd8e108561bc2ca37a13415847d87c2019f/packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js#L36-L39). no? 🧐

Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:43]: Oooh, right. Ignore my comment. :sweat:
